src/wp-includes/class-wp-scripts.php

```diff
@@ -714,6 +839,207 @@ public function in_default_dir( $src ) {
+	 * @param string $key    The data key.
+	 * @param mixed  $value  The data value.
+	 * @return bool True on success, false on failure.
+	 */
+	public function add_data( $handle, $key, $value ) {
+		if ( 'strategy' === $key ) {
+			if ( ! empty( $value ) && ! $this->is_delayed_strategy( $value ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+					/* translators: 1: $strategy, 2: $handle */
+						__( 'Invalid strategy `%1$s` defined for `%2$s` during script registration.' ),
+						$value,
+						$handle
+					),
+					'6.3.0'
+				);
+				return false;
+			} elseif ( empty( $this->registered[ $handle ]->src ) && $this->is_delayed_strategy( $value ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+						/* translators: 1: $strategy, 2: $handle */
+						__( 'Cannot supply a strategy `%1$s` for script `%2$s` because it does not have a `src` value.' ),
+						$value,
+						$handle
+					),
+					'6.3.0'
+				);
+				return false;
+			}
+		}
+		return parent::add_data( $handle, $key, $value );
+	}
+
+	/**
+	 * Checks all handles for any delayed inline scripts.
+	 *
+	 * @since 6.3.0
+	 * @see WP_Scripts::should_delay_inline_script()
+	 *
+	 * @return bool True if the inline script present, otherwise false.
+	 */
+	public function has_delayed_inline_script() {
+		foreach ( $this->registered as $handle => $script ) {
+			foreach ( array( 'before', 'after' ) as $position ) {
+				if (
+					$this->get_data( $handle, $position ) &&
+					$this->should_delay_inline_script( $handle, $position )
+				) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Gets all dependents of a script.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $handle The script handle.
+	 * @return string[] Script handles.
+	 */
+	private function get_dependents( $handle ) {
+		// Check if dependents map for the handle in question is present. If so, use it.
+		if ( array_key_exists( $handle, $this->dependents_map ) ) {
+			return $this->dependents_map[ $handle ];
+		}
+
+		$dependents = array();
+
+		// Iterate over all registered scripts, finding dependents of the script passed to this method.
+		foreach ( $this->registered as $registered_handle => $args ) {
+			if ( in_array( $handle, $args->deps, true ) ) {
+				$dependents[] = $registered_handle;
+			}
+		}
+
+		// Add the handles dependents to the map to ease future lookups.
+		$this->dependents_map[ $handle ] = $dependents;
+
+		return $dependents;
+	}
+
+	/**
+	 * Checks if the strategy passed is a valid delayed (non-blocking) strategy.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $strategy The strategy to check.
+	 * @return bool True if $strategy is one of the delayed strategies, otherwise false.
+	 */
+	private function is_delayed_strategy( $strategy ) {
+		return in_array(
+			$strategy,
+			$this->delayed_strategies,
+			true
+		);
+	}
+
+	/**
+	 * Checks if all of a script's dependents are delayed, which is required to maintain execution order.
```

Comment:
Umm, am I missing something or this is not true/wrong logic? 

> which is required to maintain execution order

Blocking scripts can **always** be used as dependencies for scripts with async and defer. I.e. the execution order is always maintained when a "delayed" script depends on a non-delayed (blocking) script. Examples:

1. `<script src="one.js">` // blocking
2. `<script async src="two.js">` // delayed

Here one.js will always be loaded and executed before two.js, and can be used as a dependency for two.js. Script Loader ensures that one.js is **always** outputted before two.js regardless of the requested "position" when one.js was registered. I.e. if one.js was registered to be loaded in the footer, but then two.js was registered to be loaded in the head and depends on one.js, one.js will also be loaded in the head before two.js.

I.e. script 1 will always be moved above/before script 2. This is the main functionality of Script Loader.

Example 2: It is also possible to ignore the output order when a `defer` script depends on a "blocking" script. This would be new functionality that may or may not be implemented at this time as it is somewhat risky (may introduce regressions for existing edge cases). However it is possible to load scripts like this:

1. `<head><script defer src="two.js"></head>` // delayed
2. `<body>...<div class="footer"><script src="one.js"></div></body>` // blocking

In this case two.js will always be executed after one.js as `defer` scripts are executed after all of the HTML is rendered (and blocking scripts are loaded and executed).

Reply:
This is a good reminder that, once it comes to documentation or a dev note, we have to clarify the terminology "dependencies vs dependents", which is confusing any time we talk about it 😆

Reply:
> This is checking to make that there are no blocking scripts that depend on the handle being checked, not the other way around.

Uh, I see, my bad :) 

> we have to clarify the terminology "dependencies vs dependents"

Yea, definitely. Also expand/enhance all the docblocks with some more understandable language even if that means to keep repeating the same explanations over and over again (should not assume everybody would know and understand the terminology that has been explained in other docblocks in core).

For example an improvement here could be like:

> Check if all scripts that depend on the current script (passed in `$handle`) have either the `defer` or `async` attribute set.